File: LLM/PromptManager.py
import base64
import os
from openai import OpenAI

# Function to encode the image
def encode_image(image_path):
    with open(image_path, "rb") as image_file:
        return base64.b64encode(image_file.read()).decode("utf-8")
    
# default structure
default_structure = {
                    "type": "json_schema",
                    "json_schema": {
                        "name": "output",
                        "schema": {
                        "type": "string",
                        "required": ["reasoning", "output"],
                        "additionalProperties": False
                        },
                        "strict": True
                    }
}

class PromptManager:
    
    def __init__(self, preprompt):
        instructions = ""
        instructions += preprompt
        
        self.dialog = [{"role": "system", "content": instructions}]
            
    def systemQuote(self, message):
        return {"role": "system", "content": message}

    def assistantQuote(self, message):
        return {"role": "assistant", "content": message}

    def userQuote(self, message):
        return {"role": "user", "content": message}

    def addQuoteToDialog(self, message, author):        
        match author.strip().lower():
            case "system":
                quote =  self.systemQuote(message)
            case "assistant":
                quote =  self.assistantQuote(message)
            case "user":
                quote = self.userQuote(message)
            case _:
                raise ValueError("Invalid author. Expected 'system', 'assistant', or 'user'.")
            
        self.dialog.append(quote)
        
    def stringAction(self, quote, structure = default_structure):
        self.dialog.append(quote)
        
        client = OpenAI(api_key = os.getenv("OPEN_API_KEY"))
        response = client.chat.completions.create(
                    model="gpt-4o-mini",
                    messages = self.dialog,
                    response_format = structure 
        )
        out_struct = response.choices[0].message
        self.addQuoteToDialog(out_struct.content, "assistant")        
        return out_struct
    
    def imageAction(self, image_path, prompt_str, structure = default_structure):
        quote = self.userQuote(prompt_str)
        
        base64_image = encode_image(image_path)
        quote["content"] = [{"type": "text","text": quote["content"]},{"type": "image_url","image_url": {"url": f"data:image/jpeg;base64,{base64_image}"}}]
        self.dialog.append(quote)
        
        client = OpenAI(api_key = os.getenv("OPEN_API_KEY"))
        response = client.chat.completions.create(
                    model="gpt-4o-mini",
                    messages = self.dialog,
                    response_format = structure
        )
        
        out_struct = response.choices[0].message
        if(not out_struct.refusal):
            self.addQuoteToDialog(out_struct.content, "assistant")        
            
        return out_struct
    

File: LLM/PromptManager.ts
import fs from 'fs';
import path from 'path';
import { OpenAI } from 'openai';
import dotenv from 'dotenv';

dotenv.config(); // Load environment variables

/**
 * Encodes an image file to a Base64 string.
 * @param imagePath Path to the image file.
 * @returns Base64 encoded string of the image.
 */
function encodeImage(imagePath: string): string {
    const imageBuffer = fs.readFileSync(imagePath);
    return imageBuffer.toString('base64');
}

/**
 * Default JSON schema structure for response formatting.
 */
const defaultStructure = {
    type: "json_schema",
    json_schema: {
        name: "output",
        schema: {
            type: "string",
            required: ["reasoning", "output"],
            additionalProperties: false
        },
        strict: true
    }
};

/**
 * Represents a single message in the conversation.
 */
type Message = {
    role: 'system' | 'assistant' | 'user';
    content: string | Array<{ type: string; text?: string; image_url?: { url: string } }>;
};

/**
 * Manages chat prompts and responses using OpenAI.
 */
class PromptManager {
    private dialog: Message[];
    private client: OpenAI;

    constructor(preprompt: string) {
        this.dialog = [{ role: "system", content: preprompt }];
        this.client = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
    }

    /**
     * Creates a system message object.
     */
    private systemQuote(message: string): Message {
        return { role: "system", content: message };
    }

    /**
     * Creates an assistant message object.
     */
    private assistantQuote(message: string): Message {
        return { role: "assistant", content: message };
    }

    /**
     * Creates a user message object.
     */
    private userQuote(message: string): Message {
        return { role: "user", content: message };
    }

    /**
     * Adds a message to the conversation history.
     * @param message The message content.
     * @param author The role of the message sender.
     */
    addQuoteToDialog(message: string, author: 'system' | 'assistant' | 'user'): void {
        const quoteMap = {
            system: this.systemQuote,
            assistant: this.assistantQuote,
            user: this.userQuote
        };

        if (!quoteMap[author]) {
            throw new Error("Invalid author. Expected 'system', 'assistant', or 'user'.");
        }

        this.dialog.push(quoteMap[author](message));
    }

    /**
     * Sends a text-based prompt to OpenAI and updates the dialog.
     * @param quote The user’s message.
     * @param structure Optional response format (default is JSON schema).
     * @returns The assistant's response.
     */
    async stringAction(quote: Message, structure = defaultStructure): Promise<Message> {
        this.dialog.push(quote);

        const response = await this.client.chat.completions.create({
            model: "gpt-4o-mini",
            messages: this.dialog,
            response_format: structure
        });

        const outStruct = response.choices[0].message as Message;
        this.addQuoteToDialog(outStruct.content as string, "assistant");
        return outStruct;
    }

    /**
     * Sends an image and text prompt to OpenAI and updates the dialog.
     * @param imagePath Path to the image file.
     * @param promptStr The user’s text prompt.
     * @param structure Optional response format (default is JSON schema).
     * @returns The assistant's response.
     */
    async imageAction(imagePath: string, promptStr: string, structure = defaultStructure): Promise<Message> {
        const base64Image = encodeImage(imagePath);
        const quote: Message = {
            role: "user",
            content: [
                { type: "text", text: promptStr },
                { type: "image_url", image_url: { url: `data:image/jpeg;base64,${base64Image}` } }
            ]
        };

        this.dialog.push(quote);

        const response = await this.client.chat.completions.create({
            model: "gpt-4o-mini",
            messages: this.dialog,
            response_format: structure
        });

        const outStruct = response.choices[0].message as Message;
        if (!(outStruct as any).refusal) {
            this.addQuoteToDialog(outStruct.content as string, "assistant");
        }

        return outStruct;
    }
}

export { PromptManager };
